### PR TITLE
Bob/custom actions

### PR DESCRIPTION
**Pull request by rjwebb**
This PR implements support for Custom Actions (see issue: https://github.com/canvasxyz/canvas/issues/132).

These are like Actions, but with some crucial differences:
- they have the `type` field set to "customAction"
- when a custom action is defined, the developer provides a JSON schema that describes the expected format of the payload. this can be any valid JSON data.
- other than checking the custom payload against the schema, the Canvas runtime does not do any validation like checking signed data etc. instead, developers will be able to write signature verification code inside the custom action definition itself, allowing contracts to support a range of signing/verification methods that do not need to be hardcoded into Canvas
- as a result of having custom payloads, we cannot guarantee that there will be a signed "timestamp" field. therefore we are setting the "timestamp" field in the model store to 0. this means that data modified by basic actions will always take precedence over data modified by custom actions.
- any model entries that are set by custom actions must have an id value that begins with the hash of the action
- the `del` database action is not permitted

## Description

Changes:
- New `CustomAction` type. This is a subclass of `Message`, alongside `Session` and `Action`, **not** a subclass of `Action`. This allowed me to implement this PR by touching as little of the existing "basic action" code as possible.
- A new table `custom_actions` in the message store for custom actions.
- Support in Core and VM for applying/executing custom actions.
- Tests for validating a spec that has custom actions.
- A test that calls a custom action and asserts that the data in the model store has been updated.

## How has this been tested?

- [x] CI tests pass
- [x] Tested with chat-next (including login, all signers, and exchanging messages)
- [x] Tested with chat-webpack (including login, all signers, and exchanging messages)
- [ ] Other:

## Does this contain any breaking changes to external interfaces?

- [ ] Hooks
- [ ] Daemon API
- [ ] Command line arguments
- [ ] Contract language

### CODE REVIEW

File: packages/core/src/rpc/utils.ts [hunk 1]
```diff
@@ -22,6 +22,8 @@ export const getMessageKey = (hash: Buffer | string, message: Message) => {
 		key.writeUintBE(message.payload.timestamp * 2 + 1, 0, 6)
 	} else if (message.type === "session") {
 		key.writeUintBE(message.payload.sessionIssued * 2, 0, 6)
+	} else if (message.type === "customAction") {
+		throw Error("getMessageKey is not yet implemented for customAction!")
```
Comment (raykyri) [inline packages/core/src/rpc/utils.ts:26]: Making a note here as a TODO - we probably want to hash custom actions to get a message key

**Comment by raykyri** [comment]
The other thing to add (either now or later) would be validating that any effects of actions write to the database with `key.startsWith(action.hash)`

File: packages/core/src/rpc/utils.ts (hunk 1 above)
Comment (rjwebb) [reply packages/core/src/rpc/utils.ts:26]: does the existing hashing approach (for session and action) assume that we never have messages that have the same timestamp? what happens if there is a collision?

Comment (rjwebb) [reply packages/core/src/rpc/utils.ts]: ok, the message key also includes the hash of the action, we can just set the timestamp to zero

**Pull request merged**